File: src/network/elixir/omni_inference_worker.ex
# OMNI Network — Elixir GenServer Inference Worker
# Concurrent inference processing with OTP supervision.

defmodule Omni.Inference.Worker do
  use GenServer
  require Logger

  defmodule State do
    defstruct [:id, :model_id, :endpoint, :stats, :status]
  end

  defmodule Stats do
    defstruct total: 0, errors: 0, total_latency_ms: 0.0

    def record(stats, latency_ms) do
      %{stats | total: stats.total + 1, total_latency_ms: stats.total_latency_ms + latency_ms}
    end

    def record_error(stats) do
      %{stats | errors: stats.errors + 1}
    end

    def avg_latency(stats) do
      if stats.total > 0, do: stats.total_latency_ms / stats.total, else: 0.0
    end
  end

  # Client API
  def start_link(opts) do
    id = Keyword.get(opts, :id, :rand.uniform(10000))
    GenServer.start_link(__MODULE__, opts, name: via(id))
  end

  def infer(worker_id, prompt, opts \\ []) do
    GenServer.call(via(worker_id), {:infer, prompt, opts}, 30_000)
  end

  def get_stats(worker_id) do
    GenServer.call(via(worker_id), :stats)
  end

  def health(worker_id) do
    GenServer.call(via(worker_id), :health)
  end

  # Server Callbacks
  @impl true
  def init(opts) do
    state = %State{
      id: Keyword.get(opts, :id, 0),
      model_id: Keyword.get(opts, :model_id, "omni-7b"),
      endpoint: Keyword.get(opts, :endpoint, "http://localhost:9090"),
      stats: %Stats{},
      status: :ready
    }
    Logger.info("Inference worker #{state.id} started (model=#{state.model_id})")
    {:ok, state}
  end

  @impl true
  def handle_call({:infer, prompt, opts}, _from, state) do
    start = System.monotonic_time(:millisecond)
    max_tokens = Keyword.get(opts, :max_tokens, 256)
    temperature = Keyword.get(opts, :temperature, 0.7)

    case do_inference(state.endpoint, prompt, max_tokens, temperature) do
      {:ok, result} ->
        latency = System.monotonic_time(:millisecond) - start
        new_stats = Stats.record(state.stats, latency)
        {:reply, {:ok, Map.put(result, :latency_ms, latency)}, %{state | stats: new_stats}}

      {:error, reason} ->
        new_stats = Stats.record_error(state.stats)
        {:reply, {:error, reason}, %{state | stats: new_stats}}
    end
  end

  @impl true
  def handle_call(:stats, _from, state) do
    result = %{
      worker_id: state.id,
      model_id: state.model_id,
      total_requests: state.stats.total,
      errors: state.stats.errors,
      avg_latency_ms: Stats.avg_latency(state.stats),
      status: state.status
    }
    {:reply, result, state}
  end

  @impl true
  def handle_call(:health, _from, state) do
    {:reply, %{status: state.status, worker_id: state.id}, state}
  end

  defp do_inference(endpoint, prompt, max_tokens, temperature) do
    body = Jason.encode!(%{
      prompt: prompt,
      max_tokens: max_tokens,
      temperature: temperature
    })

    case :httpc.request(
      :post,
      {~c"#{endpoint}/infer", [], ~c"application/json", body},
      [{:timeout, 30_000}],
      []
    ) do
      {:ok, {{_, 200, _}, _, response_body}} ->
        {:ok, Jason.decode!(to_string(response_body))}
      {:ok, {{_, code, _}, _, _}} ->
        {:error, "HTTP #{code}"}
      {:error, reason} ->
        {:error, inspect(reason)}
    end
  end

  defp via(id), do: {:via, Registry, {Omni.Inference.Registry, id}}
end

File: src/network/javascript/omni_inference_worker.js
// OMNI Concurrency Layer — JavaScript Event Loop Inference Worker
// Non-blocking inference via Web Workers and event-driven architecture.

class OmniInferenceWorker {
  constructor(config = {}) {
    this.modelEndpoint = config.endpoint || 'http://localhost:8080/api/v1';
    this.maxConcurrent = config.maxConcurrent || 10;
    this.timeout = config.timeout || 30000;
    this.activeRequests = 0;
    this.queue = [];
    this.stats = { total: 0, errors: 0, totalLatencyMs: 0 };
  }

  async infer(prompt, options = {}) {
    if (this.activeRequests >= this.maxConcurrent) {
      return new Promise((resolve, reject) => {
        this.queue.push({ prompt, options, resolve, reject });
      });
    }
    return this._executeInference(prompt, options);
  }

  async _executeInference(prompt, options) {
    this.activeRequests++;
    const start = performance.now();
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(`${this.modelEndpoint}/infer`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          prompt,
          max_tokens: options.maxTokens || 256,
          temperature: options.temperature || 0.7,
          top_p: options.topP || 0.9,
        }),
        signal: controller.signal,
      });

      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const result = await response.json();
      const latency = performance.now() - start;
      this.stats.total++;
      this.stats.totalLatencyMs += latency;
      return { ...result, latencyMs: latency };
    } catch (err) {
      this.stats.errors++;
      throw err;
    } finally {
      clearTimeout(timeoutId);
      this.activeRequests--;
      this._processQueue();
    }
  }

  async *stream(prompt, options = {}) {
    const response = await fetch(`${this.modelEndpoint}/stream`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ prompt, stream: true, ...options }),
    });

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';
      for (const line of lines) {
        if (line.startsWith('data: ') && line !== 'data: [DONE]') {
          try { yield JSON.parse(line.slice(6)); } catch {}
        }
      }
    }
  }

  _processQueue() {
    while (this.queue.length > 0 && this.activeRequests < this.maxConcurrent) {
      const { prompt, options, resolve, reject } = this.queue.shift();
      this._executeInference(prompt, options).then(resolve).catch(reject);
    }
  }

  getStats() {
    return {
      ...this.stats,
      avgLatencyMs: this.stats.total > 0 ? this.stats.totalLatencyMs / this.stats.total : 0,
      queueDepth: this.queue.length,
      activeRequests: this.activeRequests,
    };
  }
}

// Softmax utility (for client-side logit processing)
function softmax(logits) {
  const max = Math.max(...logits);
  const exps = logits.map(x => Math.exp(x - max));
  const sum = exps.reduce((a, b) => a + b, 0);
  return exps.map(x => x / sum);
}

// Cosine similarity
function cosineSimilarity(a, b) {
  let dot = 0, normA = 0, normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i]; normA += a[i] ** 2; normB += b[i] ** 2;
  }
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

if (typeof module !== 'undefined') {
  module.exports = { OmniInferenceWorker, softmax, cosineSimilarity };
}
